text/0032-dust-protection/0032-dust-protection.md

```diff
@@ -0,0 +1,126 @@
+
+A single typical `SigLockedSingleOutput` as described in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md) has a size of 42 bytes. When 1 Gi is split up into one billion outputs of 1 IOTA each, this corresponds to about 4 GB of outputs that need to be processed and stored. Since IOTA is feeless, this is far from a prohibitively expensive attack.
+
+In order to protect nodes from such attacks, one possible solution is to make accumulating dust outputs expensive. Since IOTA does not have any fees that might limit the feasibility of issuing many dust transactions, deposits pose a valid alternative to achieve a similar effect.
+
+When an address is supposed to receive micro transactions, it must have an unspent output of a special type as a deposit. This deposit cannot be consumed by any transaction as long as the dust outputs remain unspent.
+
+An additional benefit of this rule is that it makes a mass of privacy violating [forced address reuse attacks](https://en.bitcoin.it/wiki/Privacy#Forced_address_reuse) more expensive to carry out.
+
+
+# Detailed design
+
+### Definitions
+
+*Dust output*: A transaction output that has an amount smaller than 1 Mi
+
+*SigLockedDustAllowanceOutput*: A new output type for deposits that enables an address to receive dust outputs. It can be consumed as an input like a regular `SigLockedSingleOutput`.
+
+<table>
+  <tr>
+    <td><b>Name</b></td>
+    <td><b>Type</b></td>
+    <td><b>Description</b></td>
+  </tr>
+  <tr>
+    <td>Output Type</td>
+    <td>uint8</td>
+    <td>
+      Set to <strong>value 1</strong> to denote a <i>SigLockedDustAllowanceOutput</i>.
+    </td>
+  </tr>
+  <tr>
+    <td valign="top">Address</td>
+    <td colspan="2">
+      <details>
+        <summary>Ed25519 Address</summary>
+        <table>
+          <tr>
+            <td><b>Name</b></td>
+            <td><b>Type</b></td>
+            <td><b>Description</b></td>
+          </tr>
+          <tr>
+            <td>Address Type</td>
+            <td>uint8</td>
+            <td>
+              Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
+            </td>
+          </tr>
+          <tr>
+            <td>Address</td>
+            <td>Array&lt;byte&gt;[32]</td>
+            <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
+          </tr>
+        </table>
+      </details>
+    </td>
+  </tr>
+  <tr>
+    <td>Amount</td>
+    <td>uint64</td>
+    <td>The amount of tokens to deposit with this <i>SigLockedDustAllowanceOutput</i> output.</td>
+  </tr>
+</table>
+
+
+### Validation
+
+Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.
```

Comment:
We should add an optional query paramter to the API calls to decide if we want to get all UTXO on an address, or only all non-dust UTXO, or only all dust UTXO (to sweep them in the client).

This is needed for the following API calls:

```go
	// RouteAddressBech32Outputs is the route for getting all output IDs for an address.
	// The address must be encoded in bech32.
	// GET returns the outputIDs for all outputs of this address (optional query parameters: "include-spent").
	RouteAddressBech32Outputs = "/addresses/:address/outputs"

	// RouteAddressEd25519Outputs is the route for getting all output IDs for an ed25519 address.
	// The ed25519 address must be encoded in hex.
	// GET returns the outputIDs for all outputs of this address (optional query parameters: "include-spent").
	RouteAddressEd25519Outputs = "/addresses/ed25519/:address/outputs"
```

Reply:
This is kinda off-topic, and only my personal opinion, but we should add another special output type if we want to unlock an address for a colored coin.
So we don't mix the dust protection logic with the colored coin logic.
For example I may want to use colored coins as a user, but I don't want an attacker to be able to dust my address.

Reply:
- I guess there's a bigger threat from users misusing the protocol than actual attackers. It will just mean that users will have to create the dust allowance over several messages which is not so bad.
- Since dust collecting addresses should be encouraged to sweep as soon as possible I think putting this limit is a good idea.
- Definitely we should add those queries. Was waiting for that meeting with Charlie before requesting changes for the API RFC
- Maybe, not sure what the spec for colored coins will look like yet :sweat_smile: 

@Wollac, @luca-moser , @thibault-martinez , and @karimodm 
give :+1: if you agree with those modifications

Reply:
- I am all for a total upper limit, that was still an open question in the RFC. However, I've no idea whether 1000 is the right parameter here.
- I am not so sure about counting all outputs the same, no matter what there actual value is. With an upper limit, the Ti accident has already been resolved, hasn't it? To me this is just needlessly complicates the creation of dust addresses, doesn't provide any attack protection and is not really easier to implement on the node-side.

Reply:
You are right, this way it is even better.
So leave it balance based and introduce an upper limit. :+1: